### DISCUSSION TITLE
Feature Discussion: Grabbing Texture Metadata for Solid Entities

### DISCUSSION THREAD

**Discussion by RhapsodyInGeek** [func-godot/func_godot_plugin#59] [Ideas]
Interested in discussing designs for Texture MetaData for Solid Entities again. Looking for something simple that doesn't dictate design that can be implemented non-invasively and allow devs to not include the plugin in release exports if they so wish.

My thinking is that we compile the texture names into a dictionary when they're first grabbed, and when FG has finished looping through the textures, add that dictionary to the generated node's metadata. The main reason for a dictionary rather than an array is so that we can have a format of `texture name : number of faces texture is applied to`. I don't know if it's possible to get the faces, or if it would even be worth it since we don't have access to MeshInstance collision (although there might be a way to pull it off with Mesh Data Tool).

One of the more complicated issues is the separation between Mesh and Collision means that you can't simply raycast a mesh and get a specific face. It _might_ be worth it to save the averaged local positions of the vertices of the face and save that info to a PoolVector3Array. It's not perfect, but I don't know that we've actually tried that as a solution yet.

This all should probably be done within either GeoGenerator or SurfaceGatherer.

It's easy to overscope this kind of feature since we often think about it in relation to our end goals for it (material based footsteps, decals, etc...) but there are a number of different ways, different kinds of lookups and algos to deal with those features, and they have nothing to do with the actual build process, which is what FuncGodot is intended to be scoped to.

Basically, what we're looking at for ground rules:
- The feature only _provides_ data, it does **not** _send_ data
- FuncGodot must not be a Runtime requirement; its involvement must be limited to the Build process
- The data should be grabbed during a pre-existing loop
- The data needs to be saved in the generated node's metadata at the top level; it should _not_ be sent to `func_godot_properties` and it should not be sent to a child node
- The process for retrieving the data should be collision type agnostic; it should not care whether it's convex, concave, or non-existent
- The feature must be optional on a per Solid Entity basis

We also should try our best to keep the data size down. Basically, avoid nesting Arrays or Dictionaries within this top level textures Dictionary. Depending upon what we ultimately decide the data should encompass, it might be unavoidable (I'm thinking if we want to provide face positions, we'll need to use nested PoolVector3Arrays), so it's not a hard and fast rule, but definitely something we should try to maintain.

If we do face positions, then I guess it could be `"texture name" : PoolVector3Array of positions` and then if you want to find the texture counts on the Node, you can just look at the array's size.

Obviously we should recommend against using it with Worldspawn (but still allow it because we don't want to force designs on devs).

**Comment by JKarneeb** [comment]
Maybe I missed a previous discussion on this topic, but what is the desired goal of this system?

Currently, at runtime you can iterate a mesh's surfaces and corresponding materials -- although you don't have good ways to detect where those surfaces are/what faces are included.

However even if we did know to an extent, as you mention there is not a great way to do mesh level raycasts. Is the idea to come up with pre-computed information that allows you to infer something about the texture applied at a world location?

I've had success (on small scale) turning surfaces into concave polygons for collision and using that I sync world position to textures, but it's pretty heavy weight.

I think I need to better understand the goal to think competently on ways to solve it

**Comment by RhapsodyInGeek** [comment]
> at runtime you can iterate a mesh's surfaces and corresponding materials

> Is the idea to come up with pre-computed information that allows you to infer something about the texture applied at a world location?

Pretty much. Trying to optionally bake in at build time the texture info that can be used for per material / surface effects (footsteps, bullet holes, ice physics, etc...).

The goal is to gather the data for that purpose and to save it as metadata to the generated node.

**Comment by the-argus** [comment]
> One of the more complicated issues is the separation between Mesh and Collision means that you can't simply raycast a mesh and get a specific face.

I think that uniquely identifying a face is the only acceptable solution here. It's the only one that works in all cases and doesn't require level designers to think about, test, or manually configure things like footstep sounds. I have been thinking for a while that it would be a good idea to open an issue/discussion on godot and find a unified solution for concave and convex collision shapes. Of course, that could be a long and complicated process, hindered by the presence of godot-jolt and how jolt simplifies meshes after it receives them, preventing face indexing. I do think it is worth it, though. All level editors (or I guess, projects using all level editors) would benefit from such an upstream change.

And, in the meantime: I can update and alter [my PR](https://github.com/func-godot/func_godot_plugin/pull/23) to use this vertex position averaging technique you mentioned. This would unify the approach for both concave and convex collision shapes (which I believe is the only aspect of my PR which does not meet the constraints you've given, correct me if I'm wrong)

Also, about vertex position averaging: imagine a large room with short walls/ceilings and a large plane as the floor. When standing in the corner of the room, and raycasting straight down, the closest average position of any face would be one of the walls, not the floor. I think additionally including normals in the heuristic would allow users to solve that particular problem in most cases. Also, it seems a missed opportunity to not provide the user with some mapping of normal to texturename since you can uniquely identify faces in convex shapes based on normals. In the case that someone needs an accurate (but not performant) quick-and-dirty solution, they could turn on convex collisions for everything and use this normal to texturename mapping. This assumes that metadata is stored in the collision shape nodes, not the physics body nodes.

**Comment by RhapsodyInGeek** [comment]
> And, in the meantime: I can update and alter https://github.com/func-godot/func_godot_plugin/pull/23 to use this vertex position averaging technique you mentioned. This would unify the approach for both concave and convex collision shapes (which I believe is the only aspect of my PR which does not meet the constraints you've given, correct me if I'm wrong)

I think there were other issues with the PR, in that I wanted us to implement it within GeoGenerator or SurfaceGatherer due to them already iterating over the faces and accessing the texture data? I don't remember if you changed it to do the work in either of those places.

> Also, about vertex position averaging: imagine a large room with short walls/ceilings and a large plane as the floor. When standing in the corner of the room, and raycasting straight down, the closest average position of any face would be one of the walls, not the floor.

Yeah, I can see how the only solution to that would be on the LDs end, forcing them to make the floor and walls separate objects, which is not always ideal.

We might have to bite the bullet and make the Textures Dictionary a Dictionary of Dictionaries. We can't do an array, because the only reasonable way to do an Array on a complex set of data is through the use of Enums, but an Enum would create a dependency on FuncGodot at Runtime, which breaks one of the ground rules (and was also the reason we had you scrap your lookup classes).

Or maybe just an Array of Dictionaries, with each element being face data, and to get texture counts you just loop through all the Dictionaries in the array and match the Texture element, where we can keep other face info like local vertex coords, local position, and normal. We just need to remember to resize the Array and the Dictionaries so that they only take up the amount of memory they need (creating an Array or Dictionary without a size allocates an assumptive amount of memory to allow for appending new elements for performance purposes).

I'm wondering if there's any other data that would need to be grabbed?

>  This assumes that metadata is stored in the collision shape nodes, not the physics body nodes.

One of the requirements is that the Metadata would be saved to the entity Node, not the children, so it would be in the PhysicsBody nodes, not the CollisionShape nodes.

**Reply by the-argus** [reply]
What about storing the values in parallel?

```zig
{
    textures: Array[StringName]; // or PackedStringArray? both expose the count() function
    local_positions: PackedVector3Array;
    vertex_coords: PackedVector3Array;
    normals: PackedVector3Array;
}
```

Then getting texture counts is just `entity.get_meta("func_godot")["textures"].count(&"sand0tex")`

Although I am uncertain what exactly you mean by local vertex coords for each face so maybe this doesn't make sense :P

**Reply by the-argus** [reply]
Ah, additionally- could be good to have some `child_index: PackedIntArray` item here, to say what the index of the child collision shape which was generated with this face was. That was users can still correspond normals to a particular convex collision shape if they need.

**Reply by RhapsodyInGeek** [reply]
> Although I am uncertain what exactly you mean by local vertex coords for each face so maybe this doesn't make sense :P

It makes sense. When I say local vertex coords, I just mean vertex coordinates local to the MeshInstance origin as opposed to global coordinates.

While this is a good nice way to store the data, it does make the vertex coordinates a bit unwieldy to use. It's also part of Godot's style to do a number of different data structures as Dictionary results. `intersect_shape` for example returns I believe an Array of Dictionaries.`intersect_ray` is a Dictionary.

Using `StringName` instead of `String` where we can is the smart thing to do for sure.

So probably it'd be something like:
```gdscript
[
    # Tri 0
    {
        "texture": StringName("dirt"),
        "position": Vector3(0.0, 0.0, 0.0),
        "vertices": PackedVector3Array,
        "normal": Vector3(0.0, 1.0, 0.0)
    },
    # Tri 1
    {
        "texture": StringName("grass"),
        "position": Vector3(0.0, 0.0, 0.0),
        "vertices": PackedVector3Array,
        "normal": Vector3(0.0, 1.0, 0.0)
    }
]
```
That said, depending upon the complexity, that could quickly get massive. It'd probably be better to parallelize the data.

**Reply by JKarneeb** [reply]
So I may just have a very different style/approach, and if so I'll leave this to you all, however I can't help but feel like this is too abstract to be useful. 

Godot users will already have access to the raw mesh data if they want that level of detail. Finding a middle ground where something useful is extracted/generated while you are creating the mesh without literally copying the mesh data seems necessary. 

The original simple idea of: this is the list of textures used in this node, while limited, is a good example of that balance. I'd argue that a simple AABB of the surface is another example of something simple but useful and straightforward to design around. Things like average surface center while potentially useful is much harder for a designer to work around.

**Reply by RhapsodyInGeek** [reply]
> I can't help but feel like this is too abstract to be useful.

Honestly, it's far more concrete and straightforward than most data of this kind. 

The map files only have plane data, so FuncGodot has to convert intersecting planes into vertex information before it starts constructing meshes face by face. Vertex data from ArrayMesh resources are just PoolVector3Arrays. Every 3 represents the 3 vertices in a triangle.

Here we'd be either giving that data to you in a similar fashion (in Face order) or packaged up in an Array of small struct-like Dictionaries.

Having dealt with this kind of data a lot, it's incredibly useful.

> Godot users will already have access to the raw mesh data if they want that level of detail. Finding a middle ground where something useful is extracted/generated while you are creating the mesh without literally copying the mesh data seems necessary.

Generated MeshInstance nodes that contain that ArrayMesh data are children of the PhysicsBody nodes. If you're grabbing that mesh data at runtime instead of precaching it at build time, you need to perform both a `get_node("MeshInstance")` (possibly even a `if has_node("MeshInstance"):` check) and grab the MeshData _and_ loop through a PoolVector3Array of _every_ vertex in the mesh, while averaging vertex positions then doing distance checks, potentially normal checks... other stuff that I'm not thinking about right now... 

The smart thing to do is precache that data and save it to the PhysicsBody node that can actually be checked against collisions like raycasts, shapecasts... What better time to cache that information than when we're constructing it?

>  I'd argue that a simple AABB of the surface is another example of something simple but useful and straightforward to design around.

AABB of the surface would overlap with other faces, it's far too inaccurate to be useful for surface detection. Not all brushes are cubes, and not all faces are textured the same. Bad middle ground that doesn't help.

> Things like average surface center while potentially useful is much harder for a designer to work around.

We're not talking average surface center, we're talking average _polygon_ center. You have 3 points in each triangle; the averaged position of those 3 points is the center of the triangle. After FuncGodot calculates this, the data the dev receives is the 3 points _and_ the position of the polygon. This data gets saved to the PhysicsBody node.

When you raycast onto the PhysicsBody, you can then loop through each polygon position to find the closest one to the raycast collision point, maybe check if the normal of the poly is within a certain range of the raycast result's normal. Things like that.

In the absence of being able to directly raycast against a MeshInstance, it's probably the best way to try to grab individual face information (though far from perfect, but we're not left with many options due to the Collision / Mesh logic split).

The main problem isn't that the data's too abstract; it's the edge cases like @the-argus described: short walls, wide floors. That said, it's hard to know what data we have that we can provide that would help mitigate short wall, wide floor?

**Reply by JKarneeb** [reply]
Well, the last comment I'll make on the matter then, as I can't think of any additional data that you could extract that would help solve those particular issues, would be to store the information in indexed lists rather than lists of dictionaries.  I understand it makes it more complicated to find specific faces (not needing to iterate in steps of 3), but other useful Godot functions expect the data in the other form.

For example, you could create a [ConcavePolygonShape3D](https://docs.godotengine.org/en/stable/classes/class_concavepolygonshape3d.html#class-concavepolygonshape3d-method-set-faces) from the face array if you had it. This would allow someone to, if they wanted, create collision shapes for each sub-surface of their mesh relatively easily.  Granted, they could just convert the data themselves if you feel direct face interaction is the more likely interaction.

**Reply by RhapsodyInGeek** [reply]
Yeah, that's why I'm leaning more towards @the-argus 's suggestion of parallelized data rather than per face Dictionaries. Especially since PoolVector3Array works better on extremely large data sets.

**Comment by JKarneeb** [comment]
This may be a bit heavy, but seems the most similar to how I would implement this in raw Godot if I were making my own level editor.

Why not just, if opted-in, create an Area3D with collision shapes that represent each brush face extruded along its normal with info pertaining to its texture as part of the generation process? That's realistically what people are after isn't it? And it would make it quite simple to run checks against.  It just adds a relatively large addition to what gets generated during the build process.

However, doing this early on while its still at the brush stage makes it much easier to create those regions than you'll ever have again once its a mesh and the resulting objects should be very simple to use from Godot and should be performant enough if a non-default collision layer is chosen.

**Comment by RhapsodyInGeek** [comment]
> Why not just, if opted-in, create an Area3D with collision shapes that represent each brush face extruded along its normal with info pertaining to its texture as part of the generation process?

Absolutely not.

No extra nodes should come out of this process, for multiple reasons, especially ones with the performance impact that Area3Ds can have. Not to mention this violates the "avoid designing the game for devs" rule. Not every dev will use a raycast or even want the exact face that's being stood on or shot at. They may want the most common face, or the least common one, or some other information.

If someone wants to create a system where each face has a Convex CollisionShape, then the data will be there for them to do it on their own during the apply properties build step.

This feature needs to only provide data to the programmer / designer.

**Comment by sinewavey** [comment]
Something to note: I can only imagine caching this data for surfaces will take even *more* memory usage than storing information for raycast hit face index, which exists in both Jolt and Godot physics. mihe estimates face indices alone add up to an as much as 25% extra for scenes with more complex, or multiple sets of concave geometry.

Taking a look at Unreal and Unity, both offer a similar raycast-to-UV-positition function that works with (often convex only!) colliders. From early research I *think* Unreal also stores a face index buffer but I do not have access to its engine source currently. However, both seemed to indicate some kind of "mesh collider" was needed, anyway, and both also seem to indicate the additional memory requirements but in more abstract ways.

I saw many other approaches involving vertex colors, or attempts at porting the above methods into Godot. 

Personally, I think the best option will be something more akin to these methods; though how this is implemented I am not yet sure, even doubly so in a reasonably performant way. I think there is a genuine argument to be made here about a Godot feature implementation on the engine level - and not just for func_godot users.

Godot offers the Geometry3D class, which is *close* in practice, but as Tim mentions, is not fully there yet in order to not iterate over every single vertex.

**Reply by RhapsodyInGeek** [reply]
> Something to note: I can only imagine caching this data for surfaces will take even more memory usage than storing information for raycast hit face index

It's a tradeoff though, right? Do we use extra memory to precache the face data, or do we chew up more CPU performance to grab it with each collision interaction? One good "simple" example is how Sonic the Hedgehog doesn't actually do any sine or cosine calculations for angle determination: it's all prebaked into each collision tile, to keep the speed up. So instead of trying to solve for the angle, it just chooses which tile it thinks Sonic is on and gets the angle from the tile directly. More memory but faster processing, especially since that memory doesn't need to be reallocated every frame.

The truth is memory footprint minimization isn't a huge concern, since memory is cheap nowadays and these data structures are smaller than we give them credit for. I just like us being _conscious_ about it, to not be wasteful. If the data is actually useful and if we restrict their allocation to only what is needed then we're not being wasteful. That's my actual concern.

This would also be optional on a per SolidEntity basis, so devs would just have to be smart about how they use it if memory _really_ is a concern, or just opt not to use it at all.

> Personally, I think the best option will be something more akin to these methods... I think there is a genuine argument to be made here about a Godot feature implementation on the engine level - and not just for func_godot users.

The only reason I'm not entertaining the idea of creating a runtime method is because it goes out of scope for FuncGodot's mission, which is the build process. Adding methods to be used during gameplay at runtime that don't involve the build process doesn't sit well, because suddenly we're creating a dependency on FuncGodot at Runtime when part of its design is that you can remove it from your exported project by simply changing your FuncGodotMap nodes into Node3Ds if you wanted to.

So what we're limited to is providing information from the build process. That doesn't mean it's not a good idea to think about _how_ devs might grab that data from a PhysicsBody, since that informs us as to what data should be provided, but those methods aren't solutions to the FuncGodot side of things.

> Godot offers the Geometry3D class, which is close in practice, but as Tim mentions, is not fully there yet in order to not iterate over every single vertex.

So Geometry3D has this:
> *[Variant](https://docs.godotengine.org/en/stable/classes/class_variant.html#class-variant) ray_intersects_triangle(from: [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3), dir: [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3), a: [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3), b: [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3), c: [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3))* 
> *Tests if the 3D ray starting at from with the direction of dir intersects the triangle specified by a, b and c. If yes, returns the point of intersection as [Vector3](https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3). If no intersection takes place, returns null.*

If we provide the vertex data of each face with the texture assigned to it, then it should be fairly trivial for devs to use this Geometry3D method without needing to use the MeshDataTool or other methods of grabbing the face data. You basically just cache a point that's a little bit further back than the collision point (to account for fuzziness), get the direction from the new point to the collision point, and loop through the vertex array 3 vertices at a time until you get a hit, which then tells you the face index.

**Reply by sinewavey** [reply]
I do agree re: not a huge footprint. My intention with mentioning this was just to present that it is something to consider, and likely mention to end users, much the same way it's presented in Godot Jolt's documentation. This is, of course already a feature enabled by default in Godot, anyway - it's Jolt that is/was different to not include it by default.

I don't think a runtime method is something _func_godot_ should be responsible for, either, though-

To clarify, I think this is something that should be in _base Godot_ on an engine-level, regardless of using this plugin. It's something I've been tinkering with off and on in the background for a bit, since deciding on adding a similar feature to my own games with painting.

However, of course, this is a solution that goes quite far beyond what I think is reasonable to ask from the func_godot team itself.
The reason I mention it here is that I believe it's worth considering the entire picture for Godot - it wouldn't be the first, nor the last time plugins like this expose some bigger needs in the engine proper and *also* provide valid justification.

For the plugin, I do think this is probably the best we can get. I'm just a bit concerned how this plays out for complicated geometry and would be interested to see some more testing.

**Comment by RisingThumb** [comment]
It's a moot point, since the discussion seems to indicate that texture count and face information will be included, but if it's _just_ texture names, I'm against that as you can easily loop over all the materials and get the albedo texture name after a build and add it to metadata. This is what I currently do for fetching texture names present on a brush, but it does mean anything I do is per brush entity, not per face, which puts a bit more work on the level designer to work with that restriction.

I'd also ask how we're structuring the data.
```
[
    # Tri 0
    {
        "texture": StringName("dirt"),
        "position": Vector3(0.0, 0.0, 0.0),
        "vertices": PackedVector3Array,
        "normal": Vector3(0.0, 1.0, 0.0)
    },
    # Tri 1
    {
        "texture": StringName("grass"),
        "position": Vector3(0.0, 0.0, 0.0),
        "vertices": PackedVector3Array,
        "normal": Vector3(0.0, 1.0, 0.0)
    }
]
```
This structure will mean the developer has to do an array search for what they're looking for. This isn't horrible until the size of a brush entity is huge. If someone is throwing everything inadvisibly into worldspawn, then they'll have a big array to work with which will probably be a performance disaster. I'd prefer a dictionary, with the key being something that will be frequently looked up.  The key could be the texture name like below
```
{
   "grass.png": [{
        "position": Vector3(0.0, 0.0, 0.0),
        "vertices": PackedVector3Array
        "normal": Vector3(0.0, 1.0, 0.0),
    },
    {
        "position":...
    }],
    "stone.png": [{...},{...},...]
}
```
This way you can look up if a texture is present in a brush entity, you can get the count of faces with them via the array length easily, you can find which one has the most faces with one texture easily, and you still have all the face information.

Personally, I think the best would be having the key as some kind of coordinates, and some raycast or intersection test, gets that coordinates which you can then get in the dictionary and get the face information easily this way(sort of like spatial hashing). idk how hard that'd be though. A structure like this would imo be the most readily useful out of the box, as the most common use case is casting a raycast, and wanting to know what material that face is for the right decal and sfx.

Ultimately though, the structure doesn't matter _too_ much, as long as the meta data is provided before `_func_godot_apply_properties`, because at that point the developer can trim out what he doesn't need, which if all he cares about is the list of texture names, using that dictionary example, it'd just be a case of `.keys()`

**Reply by RhapsodyInGeek** [reply]
> if it's just texture names

It's not just texture names. It's providing data to the PhysicsBody that knows which faces have what textures and where those faces are.

> If someone is throwing everything inadvisibly into worldspawn...

That's their prerogative, but the default Worldspawn would have this feature disabled. I'm thinking it will be off by default for all the default entities tbh. We already tell people not to use Worldspawn for geometry in every guide just about, to only use it for quick blockouts or tests. For the most part our community has adapted to that. We can advise them not to use it on Worldspawn, then let them make that choice for themselves.

> The key could be the texture name like below

Except the goal for most folks is to find what texture you're standing on, not to check if you're standing on a specific texture. Either way, you're still having to do an array search.

The data structure I'm contemplating now would probably be more along the lines of:
```gdscript
{
    "triangles" : PoolVector3Array, # vector 3 array of all vertex positions in order
    "positions" : PoolVector3Array, # center positions of each face
    "normals" : PoolVector3Array, # normals of each face
    "textures" : Array[String] # texture applied to each face
}
```
These would be in the order of each face is in, so that each array lines up. The triangles array would naturally be 3 times the size of the other arrays but you wouldn't have to loop through each vertex, you'd still have as many iterations as the other arrays. This would allow multiple ways of looping through the data, for whatever needs you might have.

This is also how the data roughly gets structured for ArrayMesh and even other areas of FuncGodot mesh generation.

**Reply by JKarneeb** [reply]
I know that you were hoping to avoid complex nesting of data, but let me make an argument for a more complex structuring -- the following is predicated on the assumption that there is not an ability to 'lookup' a face index for a world coordinate without iterating faces (e.g., from a raycast) -- which I am not aware of existing, although I feel like @the-argus is referring to something potentially upcoming?

If instead, the structure was based around the texture as a primary key (similar to @RisingThumb above, but instead of per face, its the whole surface):
```gdscript
{
  Texture1: {
    "triangles" : PoolVector3Array, # vector 3 array of all vertex positions that use this texture (3x step)
    "positions" : PoolVector3Array, # center positions of each face that use this texture
    "normals" : PoolVector3Array, # normals of each face that use this texture
  }, Texture2: {
    ...
  } ...
}
```

I believe this has more useful properties (with one con).
1) This allows users to easily identify all textures without needing to iterate an array (its just the keys of the root dictionary)
2) It allows the user to perform checks only for certain textures
3) It allows for the creation of per-('texture')-surface collision entities easily
4) You can still iterate over all faces with the same computational complexity, although with a possible hit to memory locality

The main con of this approach, is that you lose parity with the original mesh's face index.  If somehow you had the ability to identify a face index of the original mesh, that isn't useful with my data layout as the indicies wont align.

Still, I believe that this structure will in general be more useful.

**Comment by the-argus** [comment]
sinewavey made me aware with her most recent post here that the jolt and godot-jolt devs had figured something out with the `face_index` api. Since already I was vouching for some `child_index` data for recognizing which convex shape (and therefore which set of textures) a collision corresponds to, in order to take advantage of uniquely identifying faces on convex shapes by normal, it seems like that could be extended and thought out a little more to include concave shapes.

Consider:
```zig
{
    version: String, // might be nice for implementing helper functions that operate on this metadata
    texture_names: Array[StringName];
    child_index_ranges: Array[Vector2i];

    // per-face data
    normals: PackedVector3Array;
    textures: PackedInt32Array; // index into texture_names
    local_positions: PackedVector3Array;
    vertex_coords: PackedVector3Array;
}
```

The size of each of these arrays is equal to the number of faces in the brushes on this entity. The only exceptions are `texture_names` which is just a list of the names of each of the textures on the brushes in no particular order, and `child_index_ranges` which is an array of sets of two integers which are starting index and ending index. So for concave collision or convex with only one brush, this would just be `[ Vector2i(0, num_faces) ]`. For more complex shapes, this might be something like `[ Vector2i(0, 10), Vector2i(10, 50), Vector2i(50, 55) ]` where each range is the range of faces which were generated for the child at the corresponding index. So in the latter example, the first collision shape child of this entity would be some convex shape with faces 0-10. This requires that all the faces from a given brush are all next to each other in the arrays.

You can do a lot with this data. Given the meta entry dictionary here is a variable called `meta`, here are some examples (code examples included as a sanity check, sorry if some of this stuff is obvious):

- The total number of textures applied to a brush entity would be `meta["texture_names"].size()`
- The number of faces with some texture called `ground_dirt` would be `meta["textures"].count(meta["texture_names"].find(&"ground_dirt"))`
- The texture applied to a given face would be `meta["texture_names"][meta["textures"][face_index]]` (if your raycast includes `face_index`)
- if your raycast does not include face_index, a fallback is to use the `normal` and `shape` keys:
```gdscript
func _texture_of_face_by_first_fit_normal_and_shape_idx(hit_node: Node3D, hit_normal: Vector3, shape_index: int) -> StringName:
    const MAX_COS_BETWEEN_NORMALS := 0.001
    var meta: Dictionary = hit_node.get_meta("func_godot")
    # TODO: maybe some version check here
    var range: Vector2i = meta["child_index_ranges"][shape_index]
    for idx in range(range.x, range.y):
        if hit_normal.dot(meta["normals"][idx]) < MAX_COS_BETWEEN_NORMALS:
            return meta["texture_names"][meta["textures"][idx]]
    return &""
```
- if you want a physics-agnostic approximation of the nearest face, then loop over local_positions until you find the closest one:
```gdscript
func _closest_texture_by_face_avgpos(hit_node: Node3D, hit_position: Vector3) -> StringName:
    var meta: Dictionary = hit_node.get_meta("func_godot")
    var closest_distance: float = 9999999999999999
    var closest_index: int = -1
    var idx = 0
    for local_position: Vector3 in meta["local_positions"]:
        var thisdist = hit_position.distance_to(hit_node.global_transform * local_position)
        if thisdist < closest_distance:
            closest_index = idx
            closest_distance = thisdist
        idx += 1
    return meta["texture_names"][meta["textures"][closest_index]]
```
- The above example could be altered slightly to discard faces whos `normals[idx]` value are too far from raycast normal
- If you want the most common (by # of faces) texture on the entity:
```gdscript
func _most_common_texture(hit_node: Node3D) -> StringName:
    var meta: Dictionary = hit_node.get_meta("func_godot")
    var totals_by_texture: PackedInt32Array = []
    totals_by_texture.resize(meta["texture_names"].size())
    totals_by_texture.fill(0)
    for tex_index: int in meta["textures"]:
        totals_by_texture[tex_index] += 1
    var biggest_count = -1
    var biggest_idx = -1
    var idx = 0
    for value in totals_by_texture:
        if value > biggest_count:
            biggest_count = value
            biggest_index = idx
        idx += 1
    return meta["texture_names"][biggest_idx]
```

I think this might be Good Enough (and I  think it can all be implemented in surface gatherer without adding any new loops), but here are some thoughts about potential drawbacks and a Big Solution to them:

- Users who only use one of the above methods for getting texturename for a raycast (I imagine pretty much all users) will end up with a lot of excess data.
- Maybe `child_index_ranges` is a bit too pre-designed? It could also be a bit array of child indices with a lot of repeats, leave it to the user to compress it if they need.
- If the user wants to customize this data (say, discarding the `normals`, `child_index_ranges`, `local_positions`, and `vertex_coords` if they use `face_index` for everything) then they need to find some way to visit every physics body node generated by func_godot with this metadata option enabled and perform the deletion. Make all their solid classes inherit from one common script? I don't think I'm a fan of asking that. Maybe a function in another node that runs after successful build. Oh, and maybe they don't want to delete but rather _process_ the data, in which case they need to do another loop over it, taking a potentially big hit to build time.

Okay and now for my big suggestion which should be ignored if those things aren't an issue:
All of the above issues could be solved with build-time dependency injection. If we can get expose some user-overridable object or Callable, and call a method on it, passing it the generated node, the face normals, vertices, textures, brush indices, etc, during the build process, then the user can serialize this data in their own way. Disabling the metadata would be a matter of passing an empty implementation (this empty implementation could be provided by func godot so it could be selected from a drop-down). Personally, I think this could be best solved by func_godot providing some `class_name BrushSerializer extends Resource` with some `void _func_godot_serialize_brush(out_metadata, normal, vertices_range, vertices_array, brush_idx, ...)` method that could be inherited from.

**Reply by RhapsodyInGeek** [reply]
> Consider:
> ```gdscript
> {
>     version: String, // might be nice for implementing helper functions that operate on this metadata
>     texture_names: Array[StringName];
>     child_index_ranges: Array[Vector2i];
> 
>     // per-face data
>     normals: PackedVector3Array;
>     textures: PackedInt32Array; // index into texture_names
>     local_positions: PackedVector3Array;
>     vertex_coords: PackedVector3Array;
> }
> ```

I am. I appreciate the texture look up index. While we may be holding more variables, swapping Strings for Ints and keeping a texture lookup table is definitely the right course of action.

> Maybe child_index_ranges is a bit too pre-designed? It could also be a bit array of child indices with a lot of repeats, leave it to the user to compress it if they need.

I think the main problem with it is that it assumes that child indexing is going to be constant, and doesn't account for not only offsets but manipulation. Also unsure if CollisionShapes are added before or after MeshInstances.

> Okay and now for my big suggestion which should be ignored if those things aren't an issue:
All of the above issues could be solved with...

```gdscript
## The following properties tell FuncGodot to add data to the metadata of the generated node upon build. 
## This data is parallelized, so that each element of the array is ordered to reference the same face in the mesh. 
@export_category("Mesh Metadata")

## Add a texture lookup table to the generated node's metadata on build. 
## The data is split between an [Array] of [StringName] called `"texture_names"` containing all currently used texture materials 
## and a [PackedInt32Array] called `"textures"` where each element is an index corresponding to the `"texture_names"` entries.
@export var add_textures_metadata: bool = false

## Add a [PackedVector3Array] called `"vertices"` in the generated node's metadata on build. 
## This is a list of every vertex in the generated node's [MeshInstance3D]. Every 3 vertices represent a single face.
@export var add_vertex_metadata: bool = false

## Add a [PackedVector3Array] called `"positions"` in the generated node's metadata on build. 
## This is a list of positions for each face, local to the generated node, calculated by averaging the vertices to find the face's center.
@export var add_face_position_metadata = false

## Add a [PackedVector3Array] called `"normals"` in the generated node's metadata on build. 
## Contains a list of each face's normal.
@export var add_face_normal_metadata = false
```

This does redefine the data a bit:
```gdscript
{
    "version" : String,
    "texture_names" : Array[StringName],
    "textures" : PackedInt32Array, // index into texture_names
    "vertices" : PackedVector3Array,
    "positions" : PackedVector3Array,
    "normals" : PackedVector3Array,
}
```

I'm not excluding your shape index to face range map data because I disagree, just that I think it needs a better solution given the current one assumes too much about a static child order. Maybe it should be a Dictionary where the keys are `NodePath` and the values are `Vector2i`?
```gdscript
{
    "version" : String,
    "texture_names" : Array[StringName],
    "textures" : PackedInt32Array, // index into texture_names
    "vertices" : PackedVector3Array,
    "positions" : PackedVector3Array,
    "normals" : PackedVector3Array,
    "shape_index_to_faces_range_map" : {
        "entity_1_brush_0_collision_shape" : Vector2i(0, 15),
        "entity_1_brush_1_collision_shape" : Vector2i(16, 32),
        "entity_1_brush_2_collision_shape" : Vector2i(33, 56)
    }
}
```
The only assumption this makes is that the dev isn't going to change the name of the shape after generation.

**Reply by the-argus** [reply]
Looks great to me! Nice and simple, and the shape index to faces range seems like it will work for the convex collision case. I'd like to take a stab at implementing this later this week on a new branch, if that's alright. Discussion's still open ofc though, I just want to try this solution (talking in discord, Kosmo gave me an idea for a cool way to do ice physics maybe).

**Reply by RhapsodyInGeek** [reply]
Just be patient with me while I get extremely nitpicky on how things are implemented, but by all means, if you want to tackle this feature I'm okay with that.
